app/src/main/java/org/kiwix/kiwixmobile/wifi_hotspot/HotspotService.java

```diff
@@ -0,0 +1,200 @@
+  }
+
+  @Override public int onStartCommand(Intent intent, int flags, int startId) {
+    switch (intent.getAction()) {
+
+      case ACTION_IS_HOTSPOT_ENABLED:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          serverStateListener.hotspotState(hotspotManager.checkHotspotState());
+        }
+        break;
+
+      case ACTION_TURN_ON_AFTER_O:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          //serverStateListener.hotspotTurnedOn(hotspotManager.turnOnHotspot());
+          hotspotManager.turnOnHotspot(serverStateListener);
+          updateNotification(getString(R.string.hotspot_running), true);
+        }
+        break;
+
+      case ACTION_TURN_OFF_AFTER_O:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          stopHotspot();
+        }
+        break;
+
+      case ACTION_START_SERVER:
+        webServerHelper.startServerHelper(serverStateListener);
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+          updateNotification(getString(R.string.hotspot_running), true);
+        }
+        break;
+
+      case ACTION_STOP_SERVER:
+        dismissNotification();
+        break;
+      default:
+        break;
+    }
+    return START_NOT_STICKY;
+  }
+
+  @Nullable @Override public IBinder onBind(Intent intent) {
+    return serviceBinder;
+  }
+
+  private Notification buildForegroundNotification(String status, boolean showStopButton) {
+    Log.v(TAG, "Building notification " + status);
+    builder = new NotificationCompat.Builder(this);
+    builder.setContentTitle("Kiwix Hotspot").setContentText(status);
+    Intent targetIntent = new Intent(this, ZimHostActivity.class);
+    targetIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+    PendingIntent contentIntent =
+        PendingIntent.getActivity(this, 0, targetIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+    builder.setContentIntent(contentIntent)
+        .setSmallIcon(R.mipmap.kiwix_icon)
+        .setWhen(System.currentTimeMillis());
+
+    hotspotNotificationChannel();
+
+    if (showStopButton) {
+      Intent stopIntent = new Intent(ACTION_STOP);
+      PendingIntent stopHotspot =
+          PendingIntent.getBroadcast(this, 0, stopIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+      builder.addAction(R.drawable.ic_close_white_24dp, getString(R.string.stop_hotspot_button),
+          stopHotspot);
+    }
+    return (builder.build());
+  }
+
+  private void updateNotification(String status, boolean stopAction) {
+    notificationManager.notify(HOTSPOT_NOTIFICATION_ID,
+        buildForegroundNotification(status, stopAction));
+  }
+
+  //Dismiss notification and turn off hotspot for devices>=O
+  @RequiresApi(Build.VERSION_CODES.O)
+  void stopHotspot() {
+    hotspotManager.turnOffHotspot();
+    webServerHelper.stopAndroidWebServer(serverStateListener);
+    stopForeground(true);
+    stopSelf();
+  }
+
+  //Dismiss notification and turn off hotspot for devices < O
+  void dismissNotification() {
```

Comment:
you needed this comment because this method was not named explicitly enough and also this duplicates a bunch of `stopHotspot`, you should be able to unify these methods and not have such divergent paths per api level.

```
  void stopHotspotAndCancelNotification() {
if(VersionCode>=Oreo){
      hotspotManager.turnOffHotspot();
}
    webServerHelper.stopAndroidWebServer(serverStateListener);
    stopForeground(true);
    stopSelf();
  }
```

Reply:
Fixed. a3dbdda23a2d920568ad8c010b8ea3e368e5c838